### PR TITLE
MOVING MINECARTS (by @gimilii)

### PR DESCRIPTION
**Pull request by robske110**
All credit goes to @gimilii

### CODE REVIEW

File: src/pocketmine/entity/Entity.php [hunk 1]
```diff
@@ -80,6 +80,11 @@
 	const DATA_FLAG_ACTION = 4;
 	const DATA_FLAG_INVISIBLE = 5;
 
+	const SOUTH = 0;
+	const WEST = 1;
+	const NORTH = 2;
+	const EAST = 3;
+
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Entity.php:87]: and the 45 degree?

Comment (inxomnyaa) [inline src/pocketmine/entity/Entity.php:87]: why dont u use vector3

File: src/pocketmine/entity/Minecart.php [hunk 2]
```diff
@@ -1,167 +1,423 @@
...
+use pocketmine\math\Math;
+use pocketmine\math\Vector3;
 
 class Minecart extends Vehicle{
 
-     const NETWORK_ID = 84;
+	const NETWORK_ID = 84;
 
-    public $height = 0.9;
-    public $width = 1.1;
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:14]: hey add this back

File: src/pocketmine/entity/Minecart.php [hunk 3]
```diff
@@ -1,167 +1,423 @@
...
+	const TYPE_NORMAL = 1;
+	const TYPE_CHEST = 2;
+	const TYPE_HOPPER = 3;
+	const TYPE_TNT = 4;
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php]: is there furnace?

File: src/pocketmine/entity/Minecart.php [hunk 4]
```diff
@@ -1,167 +1,423 @@
...
+	private $requestedPosition = null;
+	
+	public function initEntity(){
+		$this->setMaxHealth(1);
+		$this->setHealth($this->getMaxHealth());
+		$this->moveVector[Entity::NORTH] = new Vector3(-1, 0, 0);
+		$this->moveVector[Entity::SOUTH] = new Vector3(1, 0, 0);
+		$this->moveVector[Entity::EAST] = new Vector3(0, 0, -1);
+		$this->moveVector[Entity::WEST] = new Vector3(0, 0, 1);
+		$this->moveVector[Entity::NORTH] = new Vector3(-1, 0, 0);
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php]: These lines are duped

File: src/pocketmine/entity/Minecart.php [hunk 5]
```diff
@@ -1,167 +1,423 @@
...
+		parent::initEntity();
+	}
 
-    public function initEntity(){
-        $this->setMaxHealth(4);
-        $this->setHealth($this->getMaxHealth());
-        parent::initEntity();
-    }
+	public function getName(){
+		return "Minecart";
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:48]: fix with different types

File: src/pocketmine/entity/Minecart.php [hunk 6]
```diff
@@ -1,167 +1,423 @@
...
+			if($movingType == -1) return false;
+			elseif($movingType == 0){
+				$p = $this->getlinkedTarget();
+				if($p instanceof Player){
+					$this->motionX = -sin($p->getYaw() / 180 * M_PI);
+					$this->motionZ = cos($p->getYaw() / 180 * M_PI);
+				}
+				$target = $this->getLevel()->getBlock($this->add($this->motionX, 0, $this->motionZ)->round());
+				$target2 = $this->getLevel()->getBlock($this->add($this->motionX, 0, $this->motionZ)->floor());
+				if($target->getId() != ItemItem::AIR or $target2->getId() != ItemItem::AIR) $this->motionY = $this->gravity * 3;
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:79]: isnt it transparent

File: src/pocketmine/entity/Minecart.php [hunk 7]
```diff
@@ -1,167 +1,423 @@
...
+		return $hasUpdate or ! $this->onGround or abs ( $this->motionX ) > 0.00001 or abs ( $this->motionY ) > 0.00001 or abs ( $this->motionZ ) > 0.00001;
+	}
+	
+	/**
+	* Check if minecart is currently on a rail and if so center the cart.
+	*/
+	private function checkIfOnRail() {
+		$rail = $this->level->getBlock($this->temporalVector->setComponents($this->x, $this->y - 1, $this->z));
+		if ($rail != null && in_array($rail->getId(), [Block::RAIL, Block::ACTIVATOR_RAIL, Block::DETECTOR_RAIL, Block::POWERED_RAIL])) {
+			$rail = $rail->floor()->add(.5, 0, .5);
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:126]: -.-

File: src/pocketmine/entity/Minecart.php [hunk 8]
```diff
@@ -1,167 +1,423 @@
...
+			}
+			break;
+			case Rail::STRAIGHT_EAST_WEST :
+			switch ($candidateDirection) {
+				case Entity::WEST :
+				case Entity::EAST :
+				return $candidateDirection;
+			}
+			break;
+			case Rail::SLOPED_ASCENDING_EAST :
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:182]: fix 0.5 block Offset (move them 0.5 up)

File: src/pocketmine/entity/Minecart.php [hunk 9]
```diff
@@ -1,167 +1,423 @@
...
+				$newVector = $newVector->add(0, 2, 0);
+				// Rail could be one block up since sloping up
+				$possibleRail = $this->level->getBlock($newVector);
+				if(in_array($possibleRail->getId(), [Block::RAIL, Block::ACTIVATOR_RAIL, Block::DETECTOR_RAIL, Block::POWERED_RAIL])) {
+					$this->moveUsingVector($newVector);
+					return true;
+				} else {
+					return false;
+					// TODO Need a way to update state to say cart is stuck?
+				}
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:331]: cant move on land when pushed?

File: src/pocketmine/entity/Minecart.php (hunk 2 above)
Comment (robske110) [inline src/pocketmine/entity/Minecart.php:14]: other values are there

File: src/pocketmine/entity/Minecart.php [hunk 10]
```diff
@@ -1,167 +1,423 @@
...
+					$block = $this->level->getBlock($this->temporalVector->setComponents($x, $y, $z));
+					if(in_array($block->getId(), [Block::RAIL, Block::ACTIVATOR_RAIL, Block::DETECTOR_RAIL, Block::POWERED_RAIL])) $rails[] = $block;
+				}
+			}
+		}
 
-        return $hasUpdate or !$this->onGround or abs($this->motionX) > 0.00001 or abs($this->motionY) > 0.00001 or abs($this->motionZ) > 0.00001;
-    }*/
-    
+		$minDistance = PHP_INT_MAX;
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:377]: min = max int?!

File: src/pocketmine/entity/Minecart.php (hunk 4 above)
Comment (robske110) [inline src/pocketmine/entity/Minecart.php]: ?

File: src/pocketmine/entity/Minecart.php [hunk 11]
```diff
@@ -1,167 +1,423 @@
...
+			if($dis < $minDistance){
+				$nearestRail = $rail;
+				$minDistance = $dis;
+			}
+		}
+		return $nearestRail;
+	}
 
+	public function spawnTo(Player $player){
+		$pk = new AddEntityPacket();
```
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php]: use Player::addEntity

File: src/pocketmine/entity/Minecart.php (hunk 2 above)
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:14]: but the size is missing cant hit them properly

File: src/pocketmine/entity/Minecart.php (hunk 4 above)
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php]: dont you see you define them twice

Comment (robske110) [inline src/pocketmine/entity/Minecart.php]: oh yeah lol

File: src/pocketmine/entity/Minecart.php (hunk 3 above)
Comment (DaPigGuy) [inline src/pocketmine/entity/Minecart.php]: Dont think so.

Comment (Aericio) [inline src/pocketmine/entity/Minecart.php]: There are furnaces, in PC at least.

File: src/pocketmine/entity/Minecart.php (hunk 9 above)
Comment (robske110) [inline src/pocketmine/entity/Minecart.php:331]: yep

File: src/pocketmine/entity/Minecart.php (hunk 11 above)
Comment (robske110) [inline src/pocketmine/entity/Minecart.php]: k

File: src/pocketmine/entity/Minecart.php (hunk 3 above)
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php]: i hope they add them back, working better than b4

File: src/pocketmine/entity/Minecart.php (hunk 9 above)
Comment (inxomnyaa) [inline src/pocketmine/entity/Minecart.php:331]: how about if on land tick like normal entity?

**Pull request closed without merging**